File: conntrackReport.pl
#!/usr/bin/perl

use warnings;
use strict;

#
# TODO:
# 1. Add in argument parsing (i.e. for optional connection type header in the output)
#

my $file = '/proc/net/ip_conntrack';
my $FS = '|';	# field separator
# am I missing any TCP connection types?
my @tcpConnectionTypes = (
	"CLOSE ",	# careful not to match CLOSE_WAIT here...
	"CLOSE_WAIT",
	"ESTABLISHED",
	"FIN_WAIT",
	"FIN_WAIT2",
	"SYN_RECV",
	"TIME_WAIT"
);

my @otherConnectionTypes = (
	"udp",
	"unknown"
);

open( CONNTRACK, $file ) or die "Unable to open \'$file\' for reading: $!\n";
my @lines = <CONNTRACK>;
close CONNTRACK;

print time() . $FS;
my $connCount = @lines;	# total connections
foreach my $connectionType ( @tcpConnectionTypes ) {
	my $count = grep(/$connectionType/, @lines);
	#print $connectionType . ": " . $count . "|";
	print $count . $FS;
}

foreach my $connectionType ( @otherConnectionTypes ) {
	my $count = grep(/^$connectionType/, @lines);
	#print $connectionType . ": " . $count . "|";
	print $count . $FS;
}

print "$connCount\n";	# total 

File: conntrackReport.rb
file = '/proc/net/ip_conntrack'
FS = ','	# field separator
# am I missing any TCP connection types?
tcpConnectionTypes = [
	'CLOSE ',	# careful not to match CLOSE_WAIT here...
	'CLOSE_WAIT',
	'ESTABLISHED',
	'FIN_WAIT',
	'FIN_WAIT2',
	'SYN_RECV',
	'TIME_WAIT'
]

otherConnectionTypes = [
	'udp',
	'unknown'
]

lines = File.readlines( file )
connCount = lines.size.to_s

print Time.now.tv_sec.to_s + ','
tcpConnectionTypes.each{ |connectionType|
	print( lines.grep( /#{connectionType}/ ).size.to_s  + ',' )
}

otherConnectionTypes.each{ |connectionType|
	print( lines.grep( /#{connectionType}/ ).size.to_s + ',' )
}

puts( connCount )	# total
